pytorch_lightning/trainer/data_loading.py

```diff
@@ -119,172 +112,59 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
 
-    def auto_add_sampler(self, dataloader: Any, shuffle: bool, mode: Optional[RunningStage] = None) -> Any:
+    def _requires_distributed_sampler(self, dataloader) -> bool:
+        return (
+            self.accelerator_connector.replace_sampler_ddp
+            and self.accelerator_connector.is_distributed
+            and not isinstance(dataloader.sampler, DistributedSampler)
+            and not has_iterable_dataset(dataloader)
+        )
+
+    def prepare_dataloader(self, dataloader: Any, shuffle: bool, mode: Optional[RunningStage] = None) -> Any:
+        """This function handles to following functionalities:
+
+        - Injecting a `DistributedDataSampler` within the `DataLoader if accelertor is is_distributed
+        - Wrapping the datasets and samplers into Fault Tolerant Components
+        """
         if isinstance(dataloader, CombinedLoader):
-            # apply `auto_add_sampler` on all the collection of loaders
+            # apply `prepare_dataloader` on all the collection of loaders
             dataloader.loaders = apply_to_collection(
-                dataloader.loaders, DataLoader, self.auto_add_sampler, shuffle, mode=mode
+                dataloader.loaders, DataLoader, self.prepare_dataloader, shuffle, mode=mode
             )
             return dataloader
 
         # don't do anything if it's not a dataloader
         if not isinstance(dataloader, DataLoader):
             return dataloader
 
-        if (
-            self.accelerator_connector.replace_sampler_ddp
-            and self.accelerator_connector.is_distributed
-            and not isinstance(dataloader.sampler, DistributedSampler)
-            and not has_iterable_dataset(dataloader)
-        ):
+        if self._requires_distributed_sampler(dataloader):
             if not isinstance(dataloader.sampler, (SequentialSampler, RandomSampler)):
                 raise MisconfigurationException(
                     "You seem to have configured a sampler in your DataLoader. This will be replaced "
                     " by `DistributedSampler` since `replace_sampler_ddp` is True and you are using"
                     " distributed training. Either remove the sampler from your DataLoader or set"
                     " `replace_sampler_ddp`=False if you want to use your custom sampler."
                 )
-            sampler = self._get_distributed_sampler(dataloader, shuffle, mode=mode)
+            sampler = _get_distributed_sampler(
+                dataloader, shuffle, mode=mode, overfit_batches=self.overfit_batches, **self.distributed_sampler_kwargs
+            )
         else:
-            # use current sampler
             sampler = dataloader.sampler
 
-        dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
+        # the DataLoader should be re-created only if we need to inject
+        # the fault tolerant components or the distributed sampler.
+        if _fault_tolerant_training() or isinstance(sampler, DistributedSampler):
```

Comment:
Isn't the condition: `or isinstance(sampler, DistributedSampler)` wrong when the above if resolved to the `else` clause and the user already set a `DistributedSampler`?

We wouldn't need to call this in that case, right?

Reply:
Done !